Common/src/linear_algebra/CSysMatrix.cpp

```diff
@@ -387,96 +316,169 @@ void CSysMatrixComms::Complete(CSysVector<T>& x, CGeometry *geometry,
+}
 
-        /*--- We know the offsets based on the source rank. ---*/
+template<class T>
+void CSysMatrixComms::InitiatePeriodic(const CSysVector<T>& x, CGeometry *geometry, const CConfig *config,
+                                       const int vec_idx) {
 
-        const auto jRecv = geometry->P2PSend2Neighbor[source];
+  if (geometry->nPeriodicSend == 0) return;
 
-        /*--- Get the offset for the start of this message. ---*/
+  const auto nDim = geometry->GetnDim();
 
-        const auto msg_offset = geometry->nPoint_P2PSend[jRecv];
+  su2double rotMatrix2D[2][2] = {{1.0,0.0},{0.0,1.0}};
+  su2double rotMatrix3D[3][3] = {{1.0,0.0,0.0},{0.0,1.0,0.0},{0.0,0.0,1.0}};
 
-        /*--- Get the number of packets to be received in this message. ---*/
+  auto Rotate = [&](const su2double* direction, su2double* rotated) {
+    const su2double zeros[3] = {0.0};
+    if (nDim==2) GeometryToolbox::Rotate(rotMatrix2D, zeros, direction, rotated);
+    else GeometryToolbox::Rotate(rotMatrix3D, zeros, direction, rotated);
+  };
 
-        const auto nRecv = (geometry->nPoint_P2PSend[jRecv+1] - geometry->nPoint_P2PSend[jRecv]);
+  const unsigned short COUNT_PER_POINT = x.GetNVar();
+  const unsigned short MPI_TYPE = COMM_TYPE_DOUBLE;
 
-        SU2_OMP_FOR_STAT(CSysMatrix<T>::OMP_MIN_SIZE)
-        for (auto iRecv = 0; iRecv < nRecv; iRecv++) {
+  geometry->AllocatePeriodicComms(COUNT_PER_POINT);
 
-          /*--- Get the local index for this communicated data. ---*/
+  geometry->PostPeriodicRecvs(geometry, config, MPI_TYPE, COUNT_PER_POINT);
 
-          const auto iPoint = geometry->Local_Point_P2PSend[msg_offset + iRecv];
+  su2double* bufDSend = geometry->bufD_PeriodicSend;
 
-          /*--- Compute the offset in the recv buffer for this point. ---*/
+  for (auto iMessage = 0; iMessage < geometry->nPeriodicSend; iMessage++) {
 
-          const auto buf_offset = (msg_offset + iRecv)*COUNT_PER_POINT;
+    const auto msg_offset = geometry->nPoint_PeriodicSend[iMessage];
+    const auto nSend = geometry->nPoint_PeriodicSend[iMessage+1] - msg_offset;
 
-          /*--- Update receiving point. ---*/
+    SU2_OMP_FOR_STAT(CSysMatrix<T>::OMP_MIN_SIZE)
+    for (auto iSend = 0; iSend < nSend; iSend++) {
 
-          for (auto iVar = 0ul; iVar < x.GetNVar(); iVar++)
-            x(iPoint,iVar) += CSysMatrix<T>::template ActiveAssign<T>(bufDRecv[buf_offset+iVar]);
-        }
-        END_SU2_OMP_FOR
-        break;
+      const auto iPoint = geometry->Local_Point_PeriodicSend[msg_offset + iSend];
+      const auto buf_offset = (msg_offset + iSend)*COUNT_PER_POINT;
+
+      for (auto iVar = 0ul; iVar < x.GetNVar(); iVar++)
+        bufDSend[buf_offset+iVar] = x(iPoint,iVar);
+
+      if (vec_idx < 0) continue;
+
+      const auto iPeriodic = geometry->Local_Marker_PeriodicSend[msg_offset + iSend];
+      const auto tag = config->GetMarker_All_TagBound(iPeriodic);
+      const auto* angles = config->GetPeriodicRotAngles(tag);
```

Comment:
```suggestion
      const su2double* angles = config->GetPeriodicRotAngles(tag);
```
I think auto* for su2double* is fine :thinking:  Just resolve in case auto* works :+1: